src/main/java/de/tum/in/www1/artemis/service/LearningPathService.java

```diff
@@ -389,6 +407,322 @@ private void generateNgxGraphRepresentationForRelation(CompetencyRelation relati
+        HashMap<Long, Long> assumesCompetencies = getAssumesCompetencyMapping(learningPath.getCompetencies(), matchingClusters, priorsCompetencies);
+        Set<Long> masteredCompetencies = new HashSet<>();
+        // map of non-mastered competencies to their normalized mastery score with respect to the associated threshold
+        HashMap<Long, Double> competencyMastery = new HashMap<>();
+        learningPath.getCompetencies().forEach(competency -> {
+            // fetched learning path only contains data of the associated user
+            final var progress = competency.getUserProgress().stream().findFirst();
+            if (progress.isEmpty()) {
+                competencyMastery.put(competency.getId(), 0d);
+            }
+            else if (CompetencyProgressService.isMastered(progress.get())) {
+                // add competency to mastered set if mastered
+                masteredCompetencies.add(competency.getId());
+            }
+            else {
+                // calculate mastery progress if not completed yet
+                competencyMastery.put(competency.getId(), CompetencyProgressService.getMasteryProgress(progress.get()));
+            }
+        });
+        return new RecommendationState(masteredCompetencies, competencyMastery, matchingClusters, priorsCompetencies, extendsCompetencies, assumesCompetencies);
+    }
+
+    /**
+     * Gets a map from competency ids to a set of all other competency ids that are connected via matching relations (transitive closure, including the competency itself).
+     *
+     * @param competencies the competencies for which the mapping should be generated
+     * @return map representing the matching clusters
+     */
+    private HashMap<Long, Set<Long>> getMatchingCompetencyClusters(Set<Competency> competencies) {
+        final HashMap<Long, Set<Long>> matchingClusters = new HashMap<>();
+        for (var competency : competencies) {
+            if (!matchingClusters.containsKey(competency.getId())) {
+                final var matchingCompetencies = competencyRelationRepository.getMatchingCompetenciesByCompetencyId(competency.getId());
+                // add for each in cluster to reduce database calls (once per cluster)
+                matchingCompetencies.forEach(id -> matchingClusters.put(id, matchingCompetencies));
+            }
+        }
+        return matchingClusters;
+    }
+
+    /**
+     * Gets a map from competency ids to a set of all other competency ids that are connected via a non-matching relation.
+     *
+     * @param competencies     the competencies for which the mapping should be generated
+     * @param matchingClusters the map representing the corresponding matching clusters
+     * @return map to retrieve prior competencies
+     */
+    private HashMap<Long, Set<Long>> getPriorCompetencyMapping(Set<Competency> competencies, HashMap<Long, Set<Long>> matchingClusters) {
+        HashMap<Long, Set<Long>> priorsMap = new HashMap<>();
```

Comment:
This happens quite often in this file, I won't comment on every usage here. Please always return interface types instead of implementations.
```suggestion
    private Map<Long, Set<Long>> getPriorCompetencyMapping(Set<Competency> competencies, HashMap<Long, Set<Long>> matchingClusters) {
        Map<Long, Set<Long>> priorsMap = new HashMap<>();
```